### PR TITLE
[REVIEW] make the arena allocator work better with the producer/consumer mode

### PR DESCRIPTION
**Pull request by rongou**
In the spark-rapids plugin, when we run out of GPU memory, buffers are freed until the allocation succeeds (a.k.a. "spilling"). In this scenarios buffers are often freed from a different thread than the one originally allocated them, so we need to return the free blocks to the global arena.

Main changes to the algorithm:
* Now all allocations go through the per-thread arena, which coalesces the free blocks when possible.
* A freed "superblock" (a block with a size bigger than a threshold) is always returned to the global arena.
* When freeing a buffer that was allocated in another arena, the free block is not coalesced but returned directly to the global arena.
* The superblock size threshold is reduced, but varying this value doesn't seem to affect real query performance that much.

@harrism @jrhemstad @jlowe @abellina

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: include/rmm/mr/device/arena_memory_resource.hpp
Comment (harrism) [inline include/rmm/mr/device/arena_memory_resource.hpp:41]:
```suggestion
 * coalesced with neighbouring free blocks if the addresses are contiguous. Free superblocks are
```

File: include/rmm/mr/device/detail/arena.hpp
Comment (harrism) [inline include/rmm/mr/device/detail/arena.hpp:77]: Is this correct? If you had a superblock allocation of 1MiB, and then that block is split in half, then both would be considered superblocks even though they are part of the same upstream allocation. Deallocating the second half could fail.

File: include/rmm/mr/device/detail/arena.hpp
Comment (harrism) [inline include/rmm/mr/device/detail/arena.hpp:115]:
```suggestion
   * `this->is_contiguous_before(b)` must be true.
```

File: include/rmm/mr/device/detail/arena.hpp
Comment (harrism) [inline include/rmm/mr/device/detail/arena.hpp:416]:
```suggestion
 * An arena is a per-thread or per-non-default-stream memory pool. It allocates
```

**Review by harrism — CHANGES_REQUESTED** [review]

File: include/rmm/mr/device/detail/arena.hpp
Comment (rongou) [reply include/rmm/mr/device/detail/arena.hpp:77]: With this PR free superblocks are not kept in the per-thread arena but are always returned to the global arena, so if an allocation is bigger than the superblock threshold it's always returned when freed.

Comment (harrism) [reply include/rmm/mr/device/detail/arena.hpp:77]: Where does this happen?

Comment (rongou) [reply include/rmm/mr/device/detail/arena.hpp:77]: In `shrink_arena()`.

File: include/rmm/mr/device/detail/arena.hpp
Comment (harrism) [inline include/rmm/mr/device/detail/arena.hpp:37]: You have changed this 3 times, possibly more. Should it be an optional parameter of the class?

File: include/rmm/mr/device/detail/arena.hpp
Comment (harrism) [inline include/rmm/mr/device/detail/arena.hpp:281]: Consider using `thrust::optional` as in #601, and also use upstream->get_mem_info() for MRs that support it.

**Review by harrism — APPROVED** [review]
Approve, but a couple of suggestions.

File: include/rmm/mr/device/detail/arena.hpp
Comment (rongou) [reply include/rmm/mr/device/detail/arena.hpp:37]: It's used by both `block` and `arena`, so can't really be a non-type template parameter for either.

File: include/rmm/mr/device/detail/arena.hpp
Comment (rongou) [reply include/rmm/mr/device/detail/arena.hpp:281]: Seems like that would break existing callers? Mayer in a followup PR once #601 is merged?

Comment (harrism) [reply include/rmm/mr/device/detail/arena.hpp:281]: Yeah that's why I approved.

**Comment by rongou** [comment]
@kkraus14 and this? Thanks!

**Comment by harrism** [comment]
@kkraus14 @rongou this only had 1 C++ review. It should have had two before merging.

**Comment by rongou** [comment]
Sorry I wasn't aware of the 2 reviews rule. Is it documented somewhere? The contributing guide doesn't seem to mention it.

**Comment by harrism** [comment]
No because other teams within RAPIDS don't enforce it.  There have been countless occasions where I have seen problems pointed out by a second reviewer after being missed by the first. It also gets more people in the team familiar with more of the code in the library. I enforce this for C++ on the libraries I lead -- libcudf, librmm, libcuspatial.

Unfortunately github doesn't let us enforce 2 approvals only for C++ code owners, so instead we just have github auto-assign two C++ code owners. Once both have approved (or substitutes), it's good to merge.

**Comment by rongou** [comment]
#611

**Pull request merged**
